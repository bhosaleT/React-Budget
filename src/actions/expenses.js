import uuid from "uuid";
import database from "../firebase/firebase";

//ADD_EXPENSE
export const addExpense = expense => ({
  type: "ADD_EXPENSE",
  expense
});

export const startAddExpense = (expenseData = {}) => {
  return dispatch => {
    const {
      description = "",
      note = "",
      amount = 0,
      createdAt = 0
    } = expenseData;
    const expense = { description, note, amount, createdAt };

    return database
      .ref("expenses")
      .push(expense)
      .then(ref => {
        dispatch(
          addExpense({
            id: ref.key,
            ...expense
          })
        );
      });
  };
};

//REMOVE_EXPENSE
export const removeExpense = ({ id } = {}) => ({ type: "REMOVE_EXPENSE", id });

// EDIT_EXPEPNSE
export const editExpense = (id, updates) => ({
  type: "EDIT_EXPENSE",
  id,
  updates
});

// SET_EXPENSES
export const setExpenses = expenses => ({
  type: "SET_EXPENSES",
  expenses
});


/* HOW FETCHING DATA FROM THE DATABASE WORKS FOR THE FIRST TIME 
-- we use functions which will contact firebase and then also push our actions.
-- we will return (dispatch) to get access to dispatch
-- then we reference to database using .ref() and then .once() which then gives us a snapshot which is our data.
-- then we loop through it using forEach and get inside that we get childsnapshot we add the childsnapshot to an array 
   which we will call expenses and then just the id comes form child snapshot and then we spread the childsnapshot.val().
-- Then we will dispatch(setExpenses).
*/


// export const startSetExpenses;
export const startSetExpenses = () => {
  return dispatch => {
   return database.ref('expenses').once('value').then((snapshot)=>{
     const expenses = [];
     snapshot.forEach((childsnapshot) => {
        expenses.push({
          id: childsnapshot.key,
          ...childsnapshot.val()
        });
     });
     dispatch(setExpenses(expenses));
    })
  };
};
