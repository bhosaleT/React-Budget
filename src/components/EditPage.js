import React from "react";

const EditPage = (props) => {
    console.log(props);
    return (
        <div>
        Editing the expenses with id of {props.match.params.id}
        </div>
    )
};

export default EditPage;