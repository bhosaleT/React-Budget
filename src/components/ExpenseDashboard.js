import React from "react";

const ExpenseDashboardPage = () => <div>This is my dashboard</div>;

export default ExpenseDashboardPage;