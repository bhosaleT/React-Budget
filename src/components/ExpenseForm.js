import React, { Component } from "react";
import moment from "moment";
import { SingleDatePicker } from "react-dates";
import "react-dates/initialize";

// const date = new Date();

const now = moment();
console.log(now.format("MMM Do YYYY, h:mm a"));

export default class ExpenseForm extends Component {
  constructor(props) {
    super(props);

    this.state = {
      description: props.expense ? props.expense.description : "",
      note: props.expense ? props.expense.note : "",
      amount: props.expense ? (props.expense.amount / 100).toString() : "",
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      calFocused: false,
      error: ""
    };
  }

  onDescriptionChange = e => {
    const description = e.target.value;
    this.setState(() => ({ description }));
  };

  onChangeNote = e => {
    const note = e.target.value;
    this.setState(() => ({ note }));
  };

  onAmountChange = e => {
    const amount = e.target.value;
    //check if the amount is fitting a particular format.
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => ({ amount }));
    }
  };

  onDateChange = createdAt => {
    if (createdAt) {
      this.setState(() => ({ createdAt }));
    }
  };

  onFocusChange = ({ focused }) => {
    this.setState(() => ({ calFocused: focused }));
  };

  onSubmit = e => {
    e.preventDefault();

    if (!this.state.description || !this.state.amount) {
      this.setState(() => ({ error: "Please add a title and amount" }));
    } else {
      this.setState(() => ({ error: "" }));
      this.props.onSubmit({
        description: this.state.description,
        amount: parseFloat(this.state.amount, 10) * 100,
        createdAt: this.state.createdAt.valueOf(),
        note: this.state.note
      });
    }
  };

  render() {
    return (
      <form className="form" onSubmit={this.onSubmit}>
        {this.state.error && <p className="form__error">{this.state.error}</p>}
        <input
          type="text"
          placeholder="Title"
          autoFocus
          value={this.state.description}
          className="text-input"
          onChange={this.onDescriptionChange}
        />
        <input
          type="text"
          value={this.state.amount}
          onChange={this.onAmountChange}
          className="text-input"
          placeholder="Amount"
        />
        <SingleDatePicker
          date={this.state.createdAt}
          onDateChange={this.onDateChange}
          focused={this.state.calFocused}
          onFocusChange={this.onFocusChange}
          numberOfMonths={1}
          isOutsideRange={() => false}
        />
        <textarea
          rows="4"
          cols="40"
          className="text-area"
          placeholder="Add a note for your expense (optional)"
          value={this.state.note}
          onChange={this.onChangeNote}
        />
        <div>
          <button className="button">Add Expense</button>
        </div>
      </form>
    );
  }
}
