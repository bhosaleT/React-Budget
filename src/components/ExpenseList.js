import React from "react";
import { connect } from "react-redux";
import ExpenseListItem from "./ExpenseListItem";
import getVisibleExpenses from "../selectors/expenses";

export const ExpenseList = props => (
  <div className="content-container">
    <div className="list-header">
      <div className="show-for-mobile">Expenses</div>
      <div className="show-for-desktop">Expense</div>
      <div className="show-for-desktop">Amount</div>
    </div>

   <div className="list-body">
      {props.expenses.length === 0 ? (
        <div>
          <span className="list-item--message">No Expenses</span>
        </div>
      ) : (
          props.expenses.map(expense => {
            return <ExpenseListItem key={expense.id} {...expense} />;
            {
              /* to destructuring expenses at the ExpenseListItem we will add spreading to the props passing */
            }
          })
        )}
   </div>
  </div>
);

/* 
    ==================================== WORKING OF CONNECT =========================== 
- The connect is a named export from the package react-redux which lets us connect our react components to the redux store
- we use connect()(ExpanseList) and send back ExpanseList in it because it is a hoc which takes in a component and sends back
  a modified componet with some additions.
- we make another function which is called mapStateToProps this function takes the state from the store which is passed down from the App.js file and
  maps it to the props of the ExpanseList component.
- this can be used in props.expenses.length.
- Now our component can access all the information it requires form the store.
- As the store changes the component will change automatically.
- ********* ADDING FILTERING *********** we change the expenses props to take in getVisibleExpenses() function which takes in expenses and filters and sends back a filtered   list.
*/

const mapStateToProps = state => {
  return {
    expenses: getVisibleExpenses(state.expenses, state.filters)
  };
};

export default connect(mapStateToProps)(ExpenseList);
