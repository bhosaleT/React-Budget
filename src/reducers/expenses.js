const expensesReducerDefaultState = [];

//This reducer deals with the expenses section of the state the switch cases in this reducer will contain actions to manipulate the expenses state.
const expensesReducer = (state = expensesReducerDefaultState, action) => {
    switch (action.type) {
        case "ADD_EXPENSE":
            return [...state, action.expense];
        case "REMOVE_EXPENSE":
            return state.filter(expense => {
                return action.id !== expense.id;
            });
        case "EDIT_EXPENSE":
            return state.map(expense => {
                if (expense.id === action.id) {
                    return {
                        ...expense,
                        ...action.updates
                    };
                } else {
                    return expense;
                }
            });
        default:
            return state;
    }
};

export default expensesReducer;