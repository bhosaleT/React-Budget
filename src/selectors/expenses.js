//============================= GET VISIBLE EXPENSES ==========================//

/* 
--- WORKING OF FILTERING ---
1.  the function named getVisibleExpenses takes in two things the state.expenses and state.filters
2.  destruct the filter state into individiual variables.
3.  three variables are created to check the startDate, EndDate, text
4.  to check the startDate whether its number or not and then if expense.createdAt is less than or equal to the startDate
5.  to check if the endDate is greater than createdAt
6.  to match the text we use .includes(with the text filter provided) && convert all text to lowercase to make the filtering free or uppercase or lowercase
7.  FILTER all the values where all three are true.
8.  SortBy {using .sort() func} if the sort is based on time then check which ones createdAt timestamp value is lower that one should come first, if its sorted by amount        the larger amount one should come first.
*/

const getVisibleExpenses = (expenses, { text, sortBy, startDate, endDate }) => {
    return expenses
        .filter(expense => {
            const startDateMatch =
                typeof startDate !== "number" || expense.createdAt >= startDate;
            const endDateMatch =
                typeof endDate !== "number" || expense.createdAt <= endDate;
            const desc = expense.description.toLowerCase();
            const textMatch = desc.includes(text.toLowerCase());

            return startDateMatch && endDateMatch && textMatch;
        })
        .sort((a, b) => {
            if (sortBy === "date") {
                return a.createdAt < b.createdAt ? 1 : -1;
            } else if (sortBy === "amount") {
                return a.amount > b.amount ? -1 : 1;
            }
        });
};

export default getVisibleExpenses;