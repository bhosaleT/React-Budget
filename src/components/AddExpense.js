import React from "react";
import ExpenseForm from "./ExpenseForm";
import { connect } from "react-redux";
import { startAddExpense } from "../actions/expenses";

export class AddExpensePage extends React.Component {
  onSubmit = expense => {
    this.props.startAddExpense(expense);
    this.props.history.push("/");
  };
  render() {
    return (
      <div>
        <h1>Add Expense</h1>
        <ExpenseForm onSubmit={this.onSubmit} />
      </div>
    );
  }
}

const mapDispatchToProps = (dispatch) => {
  return {
    startAddExpense: expense => dispatch(startAddExpense(expense))
  };
};

export default connect(
  undefined,
  mapDispatchToProps
)(AddExpensePage);

/* ============= Creating a more testable component =============== 
-- during the testing phase we would have had problems tesing the addExpense because its being passed in not as a prop but imported.
-- so instead we make a new function onSubmit and to it we pass the dispatch(addExpense(expense)).
-- this is much easier to test.
*/
