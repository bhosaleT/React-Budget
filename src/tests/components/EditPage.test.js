import React from "react";
import { shallow } from "enzyme";
import expenses from "../fixtures/expenses";
import { EditPage } from "../../components/EditPage";

let editExpense, removeExpense, history, wrapper;

beforeEach(() => {
  editExpense = jest.fn();
  removeExpense = jest.fn();
  history = { push: jest.fn() };
  wrapper = shallow(
    <EditPage
      editExpense={editExpense}
      removeExpense={removeExpense}
      history={history}
      expense={expenses[2]}
    />
  );
});

//SHOULD RENDER EDITEXPENSEPAGE
test("Should render edit expense page correctly", () => {
  expect(wrapper).toMatchSnapshot();
});

test("should handle edit expense", () => {
  wrapper.find("ExpenseForm").prop("onSubmit")(expenses[2]);
  expect(history.push).toHaveBeenLastCalledWith("/");
  expect(editExpense).toHaveBeenLastCalledWith(expenses[2].id, expenses[2]);
});

test("should handle remove expense", () => {
  wrapper.find("button").simulate("click");
  expect(history.push).toHaveBeenLastCalledWith("/");
  expect(removeExpense).toHaveBeenLastCalledWith({ id: expenses[2].id });
});
