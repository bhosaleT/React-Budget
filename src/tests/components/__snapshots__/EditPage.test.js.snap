// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render edit expense page correctly 1`] = `
<div>
  <div
    className="page-header"
  >
    <div
      className="content-container"
    >
      <h1
        className="page-header__title"
      >
        Edit Expense
      </h1>
    </div>
  </div>
  <div
    className="content-container"
  >
    <ExpenseForm
      expense={
        Object {
          "amount": 1950000,
          "createdAt": 345600000,
          "description": "rent",
          "id": "3",
          "note": "",
        }
      }
      onSubmit={[Function]}
    />
    <button
      className="button"
      onClick={[Function]}
    >
      Remove
    </button>
  </div>
</div>
`;
